### DISCUSSION TITLE
Working in loop and taking orders

### DISCUSSION THREAD

**Discussion by Emmanuel789** [micropython/micropython#10954] [ESP32]
Hi guys.
I'm trying to write the program where led will be blinking in infinite loop after pushing the button and I would like my program to be able to take another orders (for example turning devices connected to different pins).
What is the best way to do it? Is it even possible?

**Comment by bixb922** [comment]
One very well structured way to do that is with asyncio, look here for a good introduction 
https://github.com/peterhinch/micropython-async/blob/master/v3/README.md

**Reply by Emmanuel789** [reply]
Thank you for your help. I will try this way.

**Comment by Emmanuel789** [comment]
Problem with this solution is that it makes other tasks only during "await" command so it's impossible to blink the led and take orders in the same time, without "await".
And another thing is that I have thread blocking calls.

**Comment by bixb922** [comment]
Yes, you can blink multiple leds and receive commands, see https://docs.micropython.org/en/latest/library/uasyncio.html.
Since asyncio is cooperative scheduling, you ```await``` as often as possible.

Blocking, threads and asyncio needs a more careful design. Here are some guidelines https://github.com/peterhinch/micropython-async/blob/master/v3/docs/THREADING.md

**Comment by Emmanuel789** [comment]
As I read there, it manages blocking calls with _thread and as far as I know it's only possible for RPI, because with ESP32 one thread is reserved for FreeRTOS only.

**Reply by bixb922** [reply]
_threads are available on the ESP32. They are scheduled time-slice fashion on the CPU dedicated to Micropython. Synchronization may be necessary to access shared variables.

**Comment by ctimmer** [comment]
This is a fairly simple polling framework that can also be used with uasyncio:
[poll-looper](https://github.com/ctimmer/poll-looper)

Curt
